core/src/main/scala/kafka/message/ByteBufferMessageSet.scala

```diff
@@ -451,20 +454,22 @@ class ByteBufferMessageSet(val buffer: ByteBuffer) extends MessageSet with Loggi
 
       if (!inPlaceAssignment) {
         // Cannot do in place assignment.
-        val wrapperMessageTimestamp = {
+        val (wrapperMessageTimestamp, offsetOfMaxTimestampInMessageSet) = {
           if (messageFormatVersion == Message.MagicValue_V0)
-            Some(Message.NoTimestamp)
+            (Some(Message.NoTimestamp), -1L)
           else if (messageFormatVersion > Message.MagicValue_V0 && messageTimestampType == TimestampType.CREATE_TIME)
-            Some(maxTimestamp)
+            (Some(maxTimestamp), {if (targetCodec == NoCompressionCodec) offsetOfMaxTimestamp else offsetCounter.value})
```

Comment:
In line 460, do we still need the test messageFormatVersion > Message.MagicValue_V0? It seems it's already covered by line 458.